File: Erlang/MergeSort.erl
-module(mergesort).
-export([mergesort/1, timeit/2, start/3]).

mergesort([])-> [];
mergesort([H]) -> [H];
mergesort(List) ->
            {Lhalf, Rhalf} = lists:split(length(List) div 2,List),
            merge(mergesort(Lhalf), mergesort(Rhalf)).


merge(Lhalf, Rhalf) ->
          merge(Lhalf, Rhalf, []).


merge([], Y, ACC) -> lists:reverse(ACC) ++ Y;
merge(X, [], ACC) -> lists:reverse(ACC) ++ X;
merge([],[], ACC) -> lists:reverse(ACC);

merge([Hl|Tl], [Hr|Tr], ACC) ->
      if Hl < Hr -> merge(Tl, [Hr|Tr], [Hl|ACC]);
        true -> merge([Hl|Tl], Tr, [Hr|ACC])
      end.


timeit(F, A) ->
        T1 = erlang:monotonic_time(),
        Val = apply(F, A),
        T2 = erlang:monotonic_time(),
        Time = erlang:convert_time_unit(T2 - T1, native, micro_seconds),
        Time.

start(FName, Counter, Result) ->
   if Counter > 0 ->
           {ok, BinaryList} = file:consult(FName),
           List = hd(BinaryList),
           NewResult = timeit(fun mergesort/1, [List]),
           start(FName, Counter-1, [NewResult|Result]);
      true -> Result
   end.

File: Haskell/MergeSort.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE BangPatterns #-}
import Control.Exception
import Formatting
import Formatting.Clock
import System.Clock
import Control.DeepSeq
import System.IO
mergesort [] = []
mergesort [x] = [x]
mergesort xs = let (lhalf, rhalf) = splitAt (length xs `div` 2) xs
               in merge' (mergesort lhalf) (mergesort rhalf)

merge' lhalf rhalf = merge lhalf rhalf []

merge [] [] acc = reverse acc
merge [] y acc = reverse acc ++ y
merge x [] acc = reverse acc ++ x

merge (l:ls) (r:rs) acc
        | l < r = merge ls (r:rs) (l:acc)
        | otherwise = merge rs (l:ls) (r:acc)

toList :: String -> [Integer]
toList input = read ("[" ++ input ++ "]")

repeater unsortedlist 0 result = return (result)

repeater unsortedlist counter result = do
                            start <- getTime Monotonic
                            let !e = mergesort unsortedlist
                            print (last e)
                            end <- getTime Monotonic
                            let newTime = toNanoSecs (diffTimeSpec start end)
                            repeater unsortedlist (counter-1) (newTime:result)

main = do
    file <- getLine
    contents <- readFile file
    let !unsortedlist = (toList contents)
    !l <- repeater unsortedlist 1000 []
    outh <- openFile "T750" WriteMode
    hPrint outh l
    hClose outh
    return ()
